File: app/src/main/java/com/esdraslopez/android/nearbychat/MessageListAdapter.java
package com.esdraslopez.android.nearbychat;

import android.app.Activity;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.util.SortedList;
import android.support.v7.widget.RecyclerView;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MessageListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int VIEW_TYPE_MESSAGE_SENT = 1;
    private static final int VIEW_TYPE_MESSAGE_RECEIVED = 2;

    private SortedList<DeviceMessage> messageList;
    private String loggedInUserUUID;
    private int selectedPosition;
    private int pressedPosition;
    private RecyclerView recyclerView;
    private final Activity activity;

    MessageListAdapter(Activity activity, String loggedInUserUUID) {
        messageList = new SortedList<>(DeviceMessage.class, new SortedList.Callback<DeviceMessage>() {
            @Override
            public int compare(DeviceMessage o1, DeviceMessage o2) {
                int creationTimeResult = Long.compare(o1.getCreationTime(), o2.getCreationTime());
                if (creationTimeResult != 0)
                    return creationTimeResult;

                return o1.getUserUUID().compareTo(o2.getUserUUID());
            }

            @Override
            public void onChanged(int position, int count) {
                notifyItemRangeChanged(position, count);
            }

            @Override
            public boolean areContentsTheSame(DeviceMessage oldItem, DeviceMessage newItem) {
                return oldItem.getMessage().equals(newItem.getMessage());
            }

            @Override
            public boolean areItemsTheSame(DeviceMessage item1, DeviceMessage item2) {
                return item1.getCreationTime() == item2.getCreationTime() &&
                        item1.getUserUUID().equals(item2.getUserUUID());
            }

            @Override
            public void onInserted(int position, int count) {
                notifyItemRangeInserted(position, count);
            }

            @Override
            public void onRemoved(int position, int count) {
                notifyItemRangeRemoved(position, count);
            }

            @Override
            public void onMoved(int fromPosition, int toPosition) {
                notifyItemMoved(fromPosition, toPosition);
            }
        });
        this.loggedInUserUUID = loggedInUserUUID;
        this.selectedPosition = RecyclerView.NO_POSITION;
        this.pressedPosition = RecyclerView.NO_POSITION;
        this.activity = activity;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view;

        switch (viewType) {
            case VIEW_TYPE_MESSAGE_SENT:
                view = LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.item_message_sent, parent, false);
                return new SentMessageHolder(view);
            case VIEW_TYPE_MESSAGE_RECEIVED:
            default:
                view = LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.item_message_received, parent, false);
                return new ReceivedMessageHolder(view);
        }
    }

    @Override
    public int getItemCount() {
        return messageList.size();
    }

    @Override
    public int getItemViewType(int position) {
        String senderUserUUID = messageList.get(position).getUserUUID();

        if (senderUserUUID.equals(loggedInUserUUID)) {
            return VIEW_TYPE_MESSAGE_SENT;
        } else {
            return VIEW_TYPE_MESSAGE_RECEIVED;
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        DeviceMessage message = messageList.get(position);

        switch (holder.getItemViewType()) {
            case VIEW_TYPE_MESSAGE_SENT:
                ((SentMessageHolder) holder).bind(message);
                break;
            case VIEW_TYPE_MESSAGE_RECEIVED:
                ((ReceivedMessageHolder) holder).bind(message);
                break;
        }
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);

        this.recyclerView = recyclerView;
    }

    public DeviceMessage get(int position) {
        return messageList.get(position);
    }

    public int add(DeviceMessage item) {
        return messageList.add(item);
    }

    public int indexOf(DeviceMessage item) {
        return messageList.indexOf(item);
    }

    public void updateItemAt(int index, DeviceMessage item) {
        messageList.updateItemAt(index, item);
    }

    public void addAll(List<DeviceMessage> items) {
        messageList.beginBatchedUpdates();
        for (DeviceMessage item : items) {
            messageList.add(item);
        }
        messageList.endBatchedUpdates();
    }

    public void addAll(DeviceMessage[] items) {
        addAll(Arrays.asList(items));
    }

    public boolean remove(DeviceMessage item) {
        return messageList.remove(item);
    }

    public DeviceMessage removeItemAt(int index) {
        return messageList.removeItemAt(index);
    }

    public void clear() {
        messageList.beginBatchedUpdates();
        while (messageList.size() > 0) {
            messageList.removeItemAt(messageList.size() - 1);
        }
        messageList.endBatchedUpdates();
    }

    abstract class MessageHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.text_message_body) TextView messageText;
        @BindView(R.id.text_message_time) TextView creationTimeText;
        ActionMode actionMode;

        MessageHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        private ActionMode.Callback actionModeCallback = new ActionMode.Callback() {

            @Override
            public boolean onCreateActionMode(ActionMode mode, Menu menu) {
                MenuInflater inflater = mode.getMenuInflater();
                inflater.inflate(R.menu.context_menu, menu);
                return true;
            }

            @Override
            public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
                return false;
            }

            @Override
            public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
                switch (item.getItemId()) {
                    case R.id.menu_share:
                        shareMessage();
                        mode.finish();
                        return true;
                    case R.id.menu_copy:
                        copyMessageToClipboard();
                        mode.finish();
                        return true;
                    default:
                        return false;
                }
            }

            @Override
            public void onDestroyActionMode(ActionMode mode) {
                actionMode = null;
                if (creationTimeText.getVisibility() != View.VISIBLE) looseFocus();
                pressedPosition = RecyclerView.NO_POSITION;
            }
        };

        void bind(DeviceMessage message) {
            messageText.setText(message.getMessageBody());
            creationTimeText.setText(Util.formatDateTime(message.getCreationTime()));

            if (selectedPosition != RecyclerView.NO_POSITION && getAdapterPosition() == selectedPosition)
                showCreationTime();
            else
                hideCreationTime();

            messageText.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {

                    if(pressedPosition != RecyclerView.NO_POSITION) {
                        MessageHolder pressedView = (MessageHolder) recyclerView.findViewHolderForAdapterPosition(pressedPosition);
                        if (pressedView != null) {
                            pressedView.actionMode.finish();
                        }
                    }

                    if (getCreationTimeText().getVisibility() == View.VISIBLE) {
                        hideCreationTime();
                        selectedPosition = RecyclerView.NO_POSITION;
                    } else {
                        hideSelectedView();
                        showCreationTime();
                        selectedPosition = getAdapterPosition();
                    }
                }
            });

            messageText.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    if (actionMode != null) {
                        return false;
                    }
                    actionMode = activity.startActionMode(actionModeCallback);
                    hideSelectedView();
                    pressedPosition = getAdapterPosition();
                    focus();
                    return true;
                }
            });
        }

        TextView getMessageText() {
            return messageText;
        }

        TextView getCreationTimeText() {
            return creationTimeText;
        }

        abstract void showCreationTime();

        abstract void hideCreationTime();

        abstract void focus();

        abstract void looseFocus();

        void shareMessage() {
            Intent sendIntent = new Intent();
            sendIntent.setAction(Intent.ACTION_SEND);
            sendIntent.putExtra(Intent.EXTRA_TEXT, getMessageText().getText());
            sendIntent.setType("text/plain");
            activity.startActivity(Intent.createChooser(sendIntent, activity.getString(R.string.send_to)));
        }

        void copyMessageToClipboard() {
            ClipboardManager clipboard = (ClipboardManager) activity.getSystemService(Context.CLIPBOARD_SERVICE);
            ClipData clip = ClipData.newPlainText("message", MessageHolder.this.getMessageText().getText());
            Objects.requireNonNull(clipboard).setPrimaryClip(clip);
            Toast.makeText(activity.getApplicationContext(), R.string.message_copied_notification, Toast.LENGTH_SHORT).show();
        }

        void hideSelectedView() {
            if (selectedPosition != RecyclerView.NO_POSITION) {
                MessageHolder selectedView = (MessageHolder) recyclerView.findViewHolderForAdapterPosition(selectedPosition);
                if (selectedView != null) {
                    selectedView.hideCreationTime();
                }
            }
        }
    }

    class SentMessageHolder extends MessageHolder {

        SentMessageHolder(View itemView) {
            super(itemView);
        }

        @Override
        void showCreationTime() {
            getCreationTimeText().setVisibility(View.VISIBLE);
            focus();
        }

        @Override
        void hideCreationTime() {
            getCreationTimeText().setVisibility(View.INVISIBLE);
            looseFocus();
        }

        @Override
        void focus() {
            getMessageText().setBackground(itemView.getResources().getDrawable(R.drawable.rounded_rectangle_primary_dark_color));
        }

        @Override
        void looseFocus() {
            getMessageText().setBackground(itemView.getResources().getDrawable(R.drawable.rounded_rectangle_primary_color));
        }
    }

    class ReceivedMessageHolder extends MessageHolder {
        @BindView(R.id.text_message_name) TextView usernameText;

        ReceivedMessageHolder(View itemView) {
            super(itemView);
        }

        void bind(DeviceMessage message) {
            super.bind(message);

            // Group consecutive messages from the same sender
            int currentPos = this.getAdapterPosition();
            if (currentPos > 0 && messageList.get(currentPos - 1).getUserUUID().equals(message.getUserUUID())) {
                usernameText.setVisibility(View.GONE);
            } else {
                usernameText.setVisibility(View.VISIBLE);
                usernameText.setText(message.getUsername());
            }
        }

        @Override
        void showCreationTime() {
            getCreationTimeText().setVisibility(View.VISIBLE);
            focus();
        }

        @Override
        void hideCreationTime() {
            getCreationTimeText().setVisibility(View.INVISIBLE);
            looseFocus();
        }

        @Override
        void focus() {
            getMessageText().setBackground(itemView.getResources().getDrawable(R.drawable.rounded_rectangle_dark_gray));
        }

        @Override
        void looseFocus() {
            getMessageText().setBackground(itemView.getResources().getDrawable(R.drawable.rounded_rectangle_light_gray));
        }
    }
}

File: app/src/main/java/com/esdraslopez/android/nearbychat/MessageListAdapter.kt
package com.esdraslopez.android.nearbychat

import android.app.Activity
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.util.SortedList
import android.support.v7.widget.RecyclerView
import android.view.*
import android.widget.TextView
import android.widget.Toast
import java.util.*

@Suppress("unused")
class MessageListAdapter internal constructor(private val activity: Activity, private val loggedInUserUUID: String) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val VIEW_TYPE_MESSAGE_SENT = 1
        private const val VIEW_TYPE_MESSAGE_RECEIVED = 2
    }

    private val messageList: SortedList<DeviceMessage>
    private var selectedPosition: Int = 0
    private var pressedPosition: Int = 0
    private lateinit var recyclerView: RecyclerView

    init {
        messageList = SortedList(DeviceMessage::class.java, object : SortedList.Callback<DeviceMessage>() {
            override fun compare(o1: DeviceMessage, o2: DeviceMessage): Int {
                val creationTimeResult = java.lang.Long.compare(o1.creationTime, o2.creationTime)
                return if (creationTimeResult != 0) creationTimeResult else o1.userUUID.compareTo(o2.userUUID)

            }

            override fun onChanged(position: Int, count: Int) {
                notifyItemRangeChanged(position, count)
            }

            override fun areContentsTheSame(oldItem: DeviceMessage, newItem: DeviceMessage): Boolean {
                return oldItem.message == newItem.message
            }

            override fun areItemsTheSame(item1: DeviceMessage, item2: DeviceMessage): Boolean {
                return item1.creationTime == item2.creationTime && item1.userUUID == item2.userUUID
            }

            override fun onInserted(position: Int, count: Int) {
                notifyItemRangeInserted(position, count)
            }

            override fun onRemoved(position: Int, count: Int) {
                notifyItemRangeRemoved(position, count)
            }

            override fun onMoved(fromPosition: Int, toPosition: Int) {
                notifyItemMoved(fromPosition, toPosition)
            }
        })
        this.selectedPosition = RecyclerView.NO_POSITION
        this.pressedPosition = RecyclerView.NO_POSITION
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view: View

        when (viewType) {
            VIEW_TYPE_MESSAGE_SENT -> {
                view = LayoutInflater.from(parent.context)
                        .inflate(R.layout.item_message_sent, parent, false)
                return SentMessageHolder(view)
            }
            VIEW_TYPE_MESSAGE_RECEIVED -> {
                view = LayoutInflater.from(parent.context)
                        .inflate(R.layout.item_message_received, parent, false)
                return ReceivedMessageHolder(view)
            }
            else -> {
                view = LayoutInflater.from(parent.context).inflate(R.layout.item_message_received, parent, false)
                return ReceivedMessageHolder(view)
            }
        }
    }

    override fun getItemCount(): Int {
        return messageList.size()
    }

    override fun getItemViewType(position: Int): Int {
        val senderUserUUID = messageList.get(position).userUUID

        return if (senderUserUUID == loggedInUserUUID) {
            VIEW_TYPE_MESSAGE_SENT
        } else {
            VIEW_TYPE_MESSAGE_RECEIVED
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val message = messageList.get(position)

        when (holder.itemViewType) {
            VIEW_TYPE_MESSAGE_SENT -> (holder as SentMessageHolder).bind(message)
            VIEW_TYPE_MESSAGE_RECEIVED -> (holder as ReceivedMessageHolder).bind(message)
        }
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)

        this.recyclerView = recyclerView
    }

    operator fun get(position: Int): DeviceMessage {
        return messageList.get(position)
    }

    fun add(item: DeviceMessage): Int {
        return messageList.add(item)
    }

    fun indexOf(item: DeviceMessage): Int {
        return messageList.indexOf(item)
    }

    fun updateItemAt(index: Int, item: DeviceMessage) {
        messageList.updateItemAt(index, item)
    }

    private fun addAll(items: List<DeviceMessage>) {
        messageList.beginBatchedUpdates()
        for (item in items) {
            messageList.add(item)
        }
        messageList.endBatchedUpdates()
    }

    fun addAll(items: Array<DeviceMessage>) {
        addAll(Arrays.asList(*items))
    }

    fun remove(item: DeviceMessage): Boolean {
        return messageList.remove(item)
    }

    fun removeItemAt(index: Int): DeviceMessage {
        return messageList.removeItemAt(index)
    }

    fun clear() {
        messageList.beginBatchedUpdates()
        while (messageList.size() > 0) {
            messageList.removeItemAt(messageList.size() - 1)
        }
        messageList.endBatchedUpdates()
    }

    internal abstract inner class MessageHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var actionMode: ActionMode? = null
        var messageBody: TextView
        var messageCreationTime: TextView

        init {
            if (this is SentMessageHolder) {
                messageBody = itemView.findViewById(R.id.sent_message_body)
                messageCreationTime = itemView.findViewById(R.id.sent_message_creation_time)
            } else {
                messageBody = itemView.findViewById(R.id.received_message_body)
                messageCreationTime = itemView.findViewById(R.id.received_message_creation_time)
            }
        }

        private val actionModeCallback = object : ActionMode.Callback {
            override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
                val inflater = mode.menuInflater
                inflater.inflate(R.menu.context_menu, menu)
                return true
            }

            override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
                return false
            }

            override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
                return when (item.itemId) {
                    R.id.menu_share -> {
                        shareMessage(messageBody.text.toString())
                        mode.finish()
                        true
                    }
                    R.id.menu_copy -> {
                        copyMessageToClipboard(messageBody.text.toString())
                        mode.finish()
                        true
                    }
                    else -> false
                }
            }

            override fun onDestroyActionMode(mode: ActionMode) {
                actionMode = null
                looseFocus()
                pressedPosition = RecyclerView.NO_POSITION
            }
        }

        internal open fun bind(message: DeviceMessage) {
            messageBody.text = message.messageBody
            messageCreationTime.text = Util.formatDateTime(message.creationTime)

            if (selectedPosition != RecyclerView.NO_POSITION && adapterPosition == selectedPosition)
                showCreationTime()
            else
                hideCreationTime()

            messageBody.setOnClickListener {
                if (pressedPosition != RecyclerView.NO_POSITION) {
                    val pressedView = recyclerView.findViewHolderForAdapterPosition(pressedPosition) as MessageHolder?
                    if (pressedView != null) {
                        pressedView.actionMode?.finish()
                    }
                }

                selectedPosition = if (messageCreationTime.visibility == View.VISIBLE) {
                    hideCreationTime()
                    RecyclerView.NO_POSITION
                } else {
                    hideSelectedView()
                    showCreationTime()
                    adapterPosition
                }
            }

            messageBody.setOnLongClickListener(View.OnLongClickListener {
                if (actionMode != null) {
                    return@OnLongClickListener false
                }
                actionMode = activity.startActionMode(actionModeCallback)
                hideSelectedView()
                pressedPosition = adapterPosition
                focus()
                true
            })
        }

        internal abstract fun showCreationTime()
        internal abstract fun hideCreationTime()
        internal abstract fun focus()
        internal abstract fun looseFocus()
    }

    internal inner class SentMessageHolder(itemView: View) : MessageHolder(itemView) {

        override fun showCreationTime() {
            messageCreationTime.visibility = View.VISIBLE
            focus()
        }

        override fun hideCreationTime() {
            messageCreationTime.visibility = View.INVISIBLE
            looseFocus()
        }

        override fun focus() {
            messageBody.background = ResourcesCompat.getDrawable(itemView.resources, R.drawable.rounded_rectangle_primary_dark_color, null)
        }

        override fun looseFocus() {
            messageBody.background = ResourcesCompat.getDrawable(itemView.resources, R.drawable.rounded_rectangle_primary_color, null)
        }
    }

    internal inner class ReceivedMessageHolder(itemView: View) : MessageHolder(itemView) {

        private var messageAuthor: TextView = itemView.findViewById(R.id.received_message_author)

        override fun bind(message: DeviceMessage) {
            super.bind(message)

            // Group consecutive messages from the same sender
            val currentPos = this.adapterPosition
            if (currentPos > 0 && messageList.get(currentPos - 1).userUUID == message.userUUID) {
                messageAuthor.visibility = View.GONE
            } else {
                messageAuthor.visibility = View.VISIBLE
                messageAuthor.text = message.username
            }
        }

        override fun showCreationTime() {
            messageCreationTime.visibility = View.VISIBLE
            focus()
        }

        override fun hideCreationTime() {
            messageCreationTime.visibility = View.INVISIBLE
            looseFocus()
        }

        override fun focus() {
            messageBody.background = ResourcesCompat.getDrawable(itemView.resources, R.drawable.rounded_rectangle_dark_gray, null)
        }

        override fun looseFocus() {
            messageBody.background = ResourcesCompat.getDrawable(itemView.resources, R.drawable.rounded_rectangle_light_gray, null)
        }
    }

    fun shareMessage(message: String) {
        val sendIntent = Intent()
        sendIntent.action = Intent.ACTION_SEND
        sendIntent.putExtra(Intent.EXTRA_TEXT, message)
        sendIntent.type = "text/plain"
        activity.startActivity(Intent.createChooser(sendIntent, activity.getString(R.string.send_to)))
    }

    fun copyMessageToClipboard(message: String) {
        val clipboard = activity.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val clip = ClipData.newPlainText("message", message)
        Objects.requireNonNull(clipboard).primaryClip = clip
        Toast.makeText(activity.applicationContext, R.string.message_copied_notification, Toast.LENGTH_SHORT).show()
    }

    fun hideSelectedView() {
        if (selectedPosition != RecyclerView.NO_POSITION) {
            val selectedView = recyclerView.findViewHolderForAdapterPosition(selectedPosition) as MessageHolder
            selectedView.hideCreationTime()
        }
    }
}